input-profile.ts

```diff
@@ -0,0 +1,280 @@
+                )) as Object3D;
+            } catch (e) {
+                console.error('failed to load 3d model');
+                console.error(e);
+                this._setComponentsActive(true);
+                console.log(
+                    'Couldnot load i-p, continuing with ' +
+                        this._handedness +
+                        ' default controller'
+                );
+            }
+            this._setComponentsActive(false);
+            this._controllerModel.parent = this.object;
+            this._controllerModel.setPositionLocal([0, 0, 0]);
+            console.log('Disabling ' + this._handedness + ' default Controller');
+            console.log(this._handedness + 'controller model loaded to the scene');
+            if (this.addVrModeSwitch)
+                this._controllerModel.addComponent(VrModeActiveSwitch);
+            this.notifyUrlReady();
+        } else {
+            console.log('mapping i-p to ' + this._handedness + ' default controllers');
+        }
+        this.getGamepadObjectsFromProfile(this._profileJSON, this._controllerModel);
+        this.setHandTrackingControllers(this.defaultController);
+        this.update = () => this.mapGamepadInput();
+    }
+
+    getGamepadObjectsFromProfile(profile: any, obj: Object3D) {
+        const components = profile.layouts[this._handedness].components;
+        if (!components) return;
+
+        for (const i in components) {
+            const visualResponses = components[i].visualResponses;
+
+            for (const j in visualResponses) {
+                // update buttons with new interface of current visual response
+                const visualResponse = visualResponses[j];
+                const valueNode = visualResponse.valueNodeName;
+                const minNode = visualResponse.minNodeName;
+                const maxNode = visualResponse.maxNodeName;
+
+                this.getGamepadObjectByName(obj, valueNode);
+                this.getGamepadObjectByName(obj, minNode);
+                this.getGamepadObjectByName(obj, maxNode);
+
+                let indice = visualResponses[j].componentProperty;
+                const response: VisualResponse = {
+                    target: this._gamepadObjects[valueNode],
+                    min: this._gamepadObjects[minNode],
+                    max: this._gamepadObjects[maxNode],
+                    id: components[i].gamepadIndices[indice], // Assign a unique ID
+                };
+                switch (indice) {
+                    case 'button':
+                        this._buttons.push(response);
+                        break;
+                    case 'xAxis':
+                        this._axes.push(response);
+                        break;
+                    case 'yAxis':
+                        this._axes.push(response);
+                        break;
+                }
+            }
+        }
+    }
+
+    getGamepadObjectByName(obj: Object3D, name: string) {
```

Comment:
I think that can be made private as well, and the name of the method is misleading. From what I understand, it searches for an object in a hiearchy and caches it.

What about also using `findByName` instead:

```ts
const found = obj.findByNameRecursive(name);
if (found[0]) this._gamepadObjects[name] = obj;
```

This is a lot less work that doing it manually?

Reply:
chnaged it so that the function  only gets object and it chaches later